File: задание 2/main.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

tuple<int, int, vector<string>> calculate_drinks(int n, int m) {
    int total_drinks = n;  // Общее количество выпитых банок
    int iterations = 0;    // Количество итераций
    int empty_bottles = n; // Количество пустых банок

    // Пошаговый процесс
    vector<string> steps;

    while (empty_bottles > 1) {
        // Вычисляем, сколько полных банок можно получить в обмен на пустые
        int new_full = empty_bottles / m;
        if (new_full == 0) { // Если банок недостаточно для обмена, прерываем цикл
            break;
        }
        // Увеличиваем общее количество выпитых банок
        total_drinks += new_full;
        // Увеличиваем количество итераций
        iterations++;
        // Вычисляем новое количество пустых банок
        empty_bottles = empty_bottles % m + new_full;
        // Записываем шаг
        steps.push_back("Step  " + to_string(iterations) + " - Drank " + to_string(new_full) + " of cans left " + to_string(empty_bottles) + " empty cans");
    }

    return make_tuple(total_drinks, iterations, steps);
}

int main() {
    int n, m;
    cout << "Enter the number of full cans: ";
    cin >> n;
    cout << "Enter the number of empty cans needed for exchange: ";
    cin >> m;

    auto result = calculate_drinks(n, m);
    int total_drinks = get<0>(result);
    int iterations = get<1>(result);
    vector<string> steps = get<2>(result);

    cout << "Maximum number of cans drunk: " << total_drinks << endl;
    cout << "Number of iterations: " << iterations << endl;
    for (const auto& step : steps) {
        cout << step << endl;
    }

    return 0;
}

File: задание 2/main.swift
func calculateDrinks(n: Int, m: Int) -> (Int, Int, [String]) {
    var totalDrinks = n
    var iterations = 0
    var emptyBottles = n
    var steps = [String]()

    while emptyBottles > 1 {
        let newFull = emptyBottles / m
        if newFull == 0 {
            break
        }
        totalDrinks += newFull
        iterations += 1
        emptyBottles = emptyBottles % m + newFull
        steps.append("Step \(iterations) - Drank \(newFull) of cans left \(emptyBottles) empty cans")
    }

    return (totalDrinks, iterations, steps)
}

let n = 15
let m = 4

let result = calculateDrinks(n: n, m: m)
let totalDrinks = result.0
let iterations = result.1
let steps = result.2

print("Maximum number of cans drunk: \(totalDrinks)")
print("Number of iterations: \(iterations)")
for step in steps {
    print(step)
}
